lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,231 @@
+                                $lastLineB = end($changedLines)->getContent();
+
+                                $startChars = StringSharedChars::startLength($firstLineA, $firstLineB);
+                                $endChars = StringSharedChars::endPos($lastLineA, $lastLineB);
+
+                                $startPos = new Position($startLineNo, $startChars);
+                                $endPos = new Position($lineNo - 1, $endChars);
+                            }
+
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range($startPos, $endPos),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $startLineNo = null;
+                            $changedLines = [];
+                        }
+
+                        if ($line->getType() === Line::UNCHANGED) {
+                            continue;
+                        }
+
+                        if ($line->getType() === Line::REMOVED) {
+                            $startLineNo = $lineNo;
+                            $changedLines[] = $line;
+
+                            continue;
+                        }
+
+                        // new line additions
+                        if ($line->getType() === Line::ADDED && $chunk->getLines()[$index - 1]->getType() === Line::UNCHANGED) {
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range(new Position($lineNo, 0), new Position($lineNo, 1)),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $contextLines = [];
+
+                            continue;
+                        }
+
+                        throw new LogicException('This is not intended to be reached');
+                    }
+                }
+            }
+
+            return $this->lastPublishedDiagnostics = $diagnostics;
+        });
+    }
+
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range, CancellationToken $cancel): Promise
+    {
+        if (empty($this->lastPublishedDiagnostics)) {
+            return new Success([]);
+        }
+
+        return \Amp\call(function () use ($textDocument) {
+            $title = 'Format with PHP CS Fixer';
+
+            $actions = [
+                CodeAction::fromArray([
+                    'title' => $title,
+                    'kind' => 'quickfix',
+                    'diagnostics' => $this->lastPublishedDiagnostics,
```

Comment:
CodeAction is added to all last published diagnostics. I don't think this should done this way. I'm not sure how else could I do this yet, but I haven't got deep into this yet.

Reply:
code action can simply recalculate :shrug: